src/JITModule.cpp

```diff
@@ -262,43 +270,84 @@ void JITModule::compile_module(std::unique_ptr<llvm::Module> m, const string &fu
 
-    TargetMachine *tm = engine_builder.selectTarget();
-    internal_assert(tm) << error_string << "\n";
-    DataLayout target_data_layout(tm->createDataLayout());
+    DataLayout target_data_layout(tm.get()->createDataLayout());
     if (initial_module_data_layout != target_data_layout) {
         internal_error << "Warning: data layout mismatch between module ("
                        << initial_module_data_layout.getStringRepresentation()
                        << ") and what the execution engine expects ("
                        << target_data_layout.getStringRepresentation() << ")\n";
     }
-    ExecutionEngine *ee = engine_builder.create(tm);
 
-    if (!ee) {
-        std::cerr << error_string << "\n";
-    }
-    internal_assert(ee) << "Couldn't create execution engine\n";
-
-    // Do any target-specific initialization
-    std::vector<llvm::JITEventListener *> listeners;
-
-    if (target.arch == Target::X86) {
-        listeners.push_back(llvm::JITEventListener::createIntelJITEventListener());
-    }
-    // TODO: If this ever works in LLVM, this would allow profiling of JIT code with symbols with oprofile.
-    // listeners.push_back(llvm::createOProfileJITEventListener());
-
-    for (auto &listener : listeners) {
-        ee->RegisterJITEventListener(listener);
+    // Create LLJIT
+    const auto compilerBuilder = [&](const llvm::orc::JITTargetMachineBuilder & /*jtmb*/)
+        -> llvm::Expected<std::unique_ptr<llvm::orc::IRCompileLayer::IRCompiler>> {
+        return std::make_unique<llvm::orc::TMOwningSimpleCompiler>(std::move(*tm));
+    };
+
+    llvm::orc::LLJITBuilderState::ObjectLinkingLayerCreator linkerBuilder;
+    if (target.arch == Target::Arch::X86 && target.bits == 32) {
+        // Fallback to RTDyld-based linking to workaround "JIT session error: Unsupported i386 relocation:4" (R_386_PLT32).
+        linkerBuilder = [&](llvm::orc::ExecutionSession &session, const llvm::Triple &) {
+            return std::make_unique<llvm::orc::RTDyldObjectLinkingLayer>(session, [&]() {
+                return std::make_unique<HalideJITMemoryManager>(dependencies);
+            });
+        };
+    } else {
+        linkerBuilder = [](llvm::orc::ExecutionSession &session, const llvm::Triple &) {
+            return std::make_unique<llvm::orc::ObjectLinkingLayer>(session);
+        };
+    }
+
+    auto JIT = llvm::cantFail(llvm::orc::LLJITBuilder()
+                                  .setDataLayout(target_data_layout)
+                                  .setCompileFunctionCreator(compilerBuilder)
+                                  .setObjectLinkingLayerCreator(linkerBuilder)
+                                  .create());
+
+    auto ctors = llvm::orc::getConstructors(*m);
+    llvm::orc::CtorDtorRunner ctorRunner(JIT->getMainJITDylib());
+    ctorRunner.add(ctors);
+
+    auto dtors = llvm::orc::getDestructors(*m);
+    llvm::orc::CtorDtorRunner *dtorRunner = new llvm::orc::CtorDtorRunner(JIT->getMainJITDylib());
```

Comment:
Should this be wrapped in `unique_ptr`? It's not clear to me where/how this gets disposed eventually.